libp2p/switch.nim

```diff
@@ -407,41 +411,63 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
 
   s.ms.addHandler(proto.codecs, proto, matcher)
 
-proc start*(s: Switch): Future[seq[Future[void]]] {.async, gcsafe.} =
-  trace "starting switch for peer", peerInfo = s.peerInfo
+proc accept(s: Switch, transport: Transport) {.async.} = # noraises
+  ## transport's accept loop
+  ##
 
-  proc handle(conn: Connection): Future[void] {.async, closure, gcsafe.} =
-    trace "Incoming connection", conn
+  while transport.running:
+    var conn: Connection
     try:
-      await s.upgradeIncoming(conn) # perform upgrade on incoming connection
-    except CancelledError as exc:
-      raise exc
+      debug "About to accept incoming connection"
+      conn = await transport.accept()
+      if not isNil(conn):
+        debug "Accepted an incoming connection", conn
+        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
+      else:
+        # A nil connection means that we might have hit a
+        # file-handle limit (or another non-fatal error),
+        # we can get one on the next try, but we should
+        # be careful to not end up in a thigh loop that
+        # will starve the main event loop, thus we sleep
+        # here before retrying.
+        await sleepAsync(100.millis) # TODO: should be configurable?
     except CatchableError as exc:
-      trace "Exception occurred in incoming handler", conn, msg = exc.msg
-    finally:
-      await conn.close()
-    trace "Connection handler done", conn
+      debug "Exception in accept loop, exiting", exc = exc.msg
+      if not isNil(conn):
+        await conn.closeWithEOF()
```

Comment:
closeWithEOF isn't really appropriate here since we don't know if the other end will be sending an EOF really - ie the connection should just be `close`:ed - if it were an LPChannel, it would need a `reset` - probably, the concept of half-closed connections needs to be moved to `LPStream` itself so that user code can distinguish between `close` and `reset` (which have unfortunate names, given decades of `shutdown` and `close` being used for this purpose)